osquery/tables/system/linux/md_tables.cpp

```diff
@@ -0,0 +1,771 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+#include <errno.h>
+#include <fcntl.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+
+#include <linux/major.h>
+#include <linux/raid/md_p.h>
+#include <linux/raid/md_u.h>
+
+#include <fstream>
+#include <memory>
+#include <numeric>
+
+#include <boost/algorithm/string/trim.hpp>
+
+#include <osquery/core/conversions.h>
+#include <osquery/filesystem.h>
+#include <osquery/logger.h>
+
+#include "osquery/events/linux/udev.h"
+#include "osquery/tables/system/linux/md_tables.h"
+
+namespace osquery {
+namespace tables {
+
+std::string kMDStatPath = "/proc/mdstat";
+
+/**
+ * @brief Removes prefixing and suffixing character from each string in vector
+ *
+ * @param strs reference to vector of target strings
+ *
+ */
+static inline void trimStrs(std::vector<std::string>& strs) {
+  for (auto& s : strs) {
+    boost::algorithm::trim(s);
+  }
+}
+
+/**
+ * @brief reads mdstat file and parses as vector of lines
+ *
+ * @param lines reference to vector of strings to store lines in
+ *
+ * Trims all white space in lines and ignores empty lines (i.e. only spaces,
+ * tabs, etc.)
+ */
+static inline void getLines(std::vector<std::string>& lines) {
+  Status status(pathExists(kMDStatPath));
+  if (status.getCode() != 0) {
+    return;
+  }
+
+  std::string content("");
+  status = readFile(kMDStatPath, content);
+  if (status.getCode() != 0) {
+    return;
+  }
+
+  lines = split(content, "\n");
+  trimStrs(lines);
+}
+
+/**
+ * @brief function for walking thru a udev subsystem and working on devices
+ *
+ * @param systemName the name of the sysfs subsytem to work with
+ * @param f function to execute on the subsystem; returns true if can break the
+ * device loop, false to keep walking
+ *
+ */
+void walkUdevDevices(const std::string& systemName,
+                     std::function<bool(udev_device* const&)> f) {
+  auto delUdev = [](udev* u) { udev_unref(u); };
+  std::unique_ptr<udev, decltype(delUdev)> handle(udev_new(), delUdev);
+  if (handle.get() == nullptr) {
+    LOG(ERROR) << "Could not get udev handle\n";
```

Comment:
Whops sorry, forgot to remove those (did initial development of this table outside of the osquery project space)